From f686655a2369bef12f09829abb07f8f33d9f9aa8 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Tue, 3 Jan 2017 17:26:42 +0800
Subject: [PATCH 1449/5242] MLK-13714 video: mxsfb: fix the overlay wrong
 offset issue

commit  94f4b9f25a0e4f1d25087f91b02bf0ac12a3a5f7 from
https://source.codeaurora.org/external/imx/linux-imx.git

The overlay function should be enabled when the lcdif is disabled
to make them synchronous. Otherwise, there will be offset for the
overlay surface when display.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index 2c35219..c113592 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -1556,11 +1556,23 @@ static void overlayfb_enable(struct mxsfb_layer *ofb)
 	uint32_t as_ctrl = 0x0;
 	struct mxsfb_info *fbi = ofb->fbi;
 
+	lock_fb_info(fbi->fb_info);
+	if (fbi->cur_blank == FB_BLANK_UNBLANK) {
+		mxsfb_disable_controller(fbi->fb_info);
+		writel(CTRL1_FIFO_CLEAR, fbi->base + LCDC_CTRL1 + REG_SET);
+	}
+
 	/* enable AS */
 	as_ctrl = readl(fbi->base + LCDC_AS_CTRL);
 	as_ctrl |= 0x1;
 
 	writel(as_ctrl, fbi->base + LCDC_AS_CTRL);
+
+	if (fbi->cur_blank == FB_BLANK_UNBLANK) {
+		writel(CTRL1_FIFO_CLEAR, fbi->base + LCDC_CTRL1 + REG_CLR);
+		mxsfb_enable_controller(fbi->fb_info);
+	}
+	unlock_fb_info(fbi->fb_info);
 }
 
 static void overlayfb_disable(struct mxsfb_layer *ofb)
@@ -1568,10 +1580,8 @@ static void overlayfb_disable(struct mxsfb_layer *ofb)
 	uint32_t as_ctrl = 0x0;
 	struct mxsfb_info *fbi = ofb->fbi;
 
-	as_ctrl = readl(fbi->base + LCDC_AS_CTRL);
-	as_ctrl &= 0xfffffffe;
-
 	writel(as_ctrl, fbi->base + LCDC_AS_CTRL);
+	writel(0x0, fbi->base + LCDC_AS_NEXT_BUF);
 }
 
 static void overlayfb_setup(struct mxsfb_layer *ofb)
-- 
1.7.9.5

